src/serac/physics/utilities/finite_element_state.hpp

```diff
@@ -241,23 +247,182 @@ class FiniteElementState {
+   */
+  QuadratureData(mfem::Mesh& mesh, const int p, const bool alloc = true);
+
+  QuadratureData(mfem::QuadratureFunction& qfunc)
+      : qspace_(qfunc.GetSpace()), qfunc_(&qfunc), data_(static_cast<std::size_t>(qfunc.Size() / stride_))
+  {
+    const double* qfunc_ptr = detail::retrieve(qfunc_).GetData();
+    int           j         = 0;
+    T*            data_ptr  = data_.data();
+    for (int i = 0; i < detail::retrieve(qfunc_).Size(); i += stride_) {
+      // The only legal (portable, defined) way to do type punning in C++
+      std::memcpy(data_ptr + j, qfunc_ptr + i, sizeof(T));
+      j++;
+    }
+  }
+
+  /**
+   * @brief Retrieves the data for a given quadrature point
+   * @param[in] element_idx The index of the desired element within the mesh
+   * @param[in] q_idx The index of the desired quadrature point within the element
+   */
+  T& operator()(const int element_idx, const int q_idx);
+
+  /**
+   * @brief Assigns an item to each quadrature point
+   * @param[in] item The item to assign
+   */
+  QuadratureData& operator=(const T& item);
+
+  /**
+   * @brief Iterator to the data for the first quadrature point
+   */
+  auto begin() { return data_.begin(); }
+  /// @overload
+  auto begin() const { return data_.begin(); }
+  /**
+   * @brief Iterator to one element past the data for the last quadrature point
+   */
+  auto end() { return data_.end(); }
+  /// @overload
+  auto end() const { return data_.end(); }
+
+  mfem::QuadratureFunction& QFunc() { return detail::retrieve(qfunc_); }
+
+  /**
+   * @brief Synchronizes data from the stored vector<T> to the raw double*
+   * array used by the underlying mfem::QuadratureFunction
+   *
+   * Used for saving to a file - MFEMSidreDataCollection
+   * (and by extension mfem::DataCollection's interface) only allow for
+   * quadrature-point-specific data via mfem::QuadratureFunction, so this logic
+   * is needed to glue together a generic array of data with that class
+   */
+  void sync() override
+  {
+    double*  qfunc_ptr = detail::retrieve(qfunc_).GetData();
+    int      j         = 0;
+    const T* data_ptr  = data_.data();
+    for (int i = 0; i < detail::retrieve(qfunc_).Size(); i += stride_) {
+      // The only legal (portable, defined) way to do type punning in C++
+      std::memcpy(qfunc_ptr + i, data_ptr + j, sizeof(T));
+      j++;
+    }
+  }
+
+private:
+  // FIXME: These will probably need to be MaybeOwningPointers
+  // See https://github.com/LLNL/axom/pull/433
+  /**
+   * @brief Storage layout of @p qfunc_ containing mesh and polynomial order info
+   */
+  detail::MaybeOwningPointer<mfem::QuadratureSpace> qspace_;
+  /**
+   * @brief Per-quadrature point data, stored as array of doubles for compatibility with Sidre
+   */
+  detail::MaybeOwningPointer<mfem::QuadratureFunction> qfunc_;
+
+  std::vector<T> data_;
+  /**
+   * @brief The stride of the array
+   */
+  static constexpr int stride_ = sizeof(T) / sizeof(double);
```

Comment:
if `T` is `int`, wouldn't `stride_` be 0? More generally, if `sizeof(T)` isn't divisible by 8, this seems like it gives the wrong value (or I am misinterpreting the meaning of `stride`).

```suggestion
  static constexpr int stride_ = (sizeof(T) + sizeof(double) - 1) / sizeof(double);
```
https://godbolt.org/z/34KvKoao9

Reply:
You are completely correct - I think semantically what we want to do is have `stride_` be of floating-point type but that opens another can of worms - with sizes (for example) we'd have to make sure to always round up, and there's also some pointer arithmetic logic that would have to be refactored somehow

Reply:
another option is to use bresenham's algorithm to handle the case of when the sizes don't divide nicely. In theory, there's some optimizations to be made for when `sizeof(T) < 8`, but I'm not sure it's worth worrying about too much (in practice the user data will almost always be greater than 8 bytes).